### PR TITLE
feat: kuberpult initial sqlite local implementation

### PR DESCRIPTION
**Pull request by miguel-crespo-fdc**
Cd-service runs a migration script on startup. Some dummy content is inserted into a dummy table and simply logged for now.

Story ID: SRX-SH558L

### CODE REVIEW

File: services/cd-service/pkg/repository/db.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	"github.com/golang-migrate/migrate/v4/source/file"
+	"github.com/golang-migrate/migrate/v4/source/iofs"
+	"go.uber.org/zap"
+	"math/rand"
+	"path"
+	"time"
+)
+
+func GetDBConnection(dbFolderLocation string) (*sql.DB, error) {
+	return sql.Open("sqlite3", path.Join(dbFolderLocation, "db.sqlite")) //not clear on what is needed for the user and password
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/db.go:36]:
```suggestion
	return sql.Open("sqlite3", path.Join(dbFolderLocation, "db.sqlite"))
```

We can ignore the db authentication as long as we're just working with sqlite

File: services/cd-service/pkg/repository/db.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	}
+	defer db.Close()
+
+	driver, err := sqlite.WithInstance(db, &sqlite.Config{
+		DatabaseName:    "",
+		MigrationsTable: "",
+		NoTxWrap:        false,
+	})
+	if err != nil {
+		logger.FromContext(ctx).Warn("Error creating DB driver. Error: ", zap.Error(err))
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/db.go:53]: same here about error handling

File: services/cd-service/pkg/repository/db.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		return err
+	}
+
+	migrationsSrc, err := (&file.File{PartialDriver: iofs.PartialDriver{}}).Open(path.Join(dbFolderLocation, "migrations"))
+	if err != nil {
+		logger.FromContext(ctx).Warn("Error opening DB migrations. Error: ", zap.Error(err))
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/db.go:59]: same here about error handling

File: services/cd-service/pkg/repository/db.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+		return err
+	}
+	m, err := migrate.NewWithInstance("file", migrationsSrc, "sqlite3", driver)
+	if err != nil {
+		logger.FromContext(ctx).Warn("Error creating migration instance. Error: ", zap.Error(err))
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/db.go:64]: same here about error handling

File: services/cd-service/pkg/repository/db_test.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+	for _, tc := range tcs {
+		tc := tc
+		t.Run(tc.Name, func(t *testing.T) {
+			t.Parallel()
+			dir := t.TempDir()
+			connection, err := GetDBConnection(dir)
+			if err != nil {
+				t.Fatalf("Error establishing DB connection. Error: %v\n", err)
+			}
+			pingErr := connection.Ping()
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/db_test.go:52]:
```suggestion
                        defer connection.Close()
			pingErr := connection.Ping()
```

File: services/cd-service/pkg/repository/db_test.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+			m := map[int]dummyDbRow{}
+			for rows.Next() {
+				r := dummyDbRow{}
+				err := rows.Scan(&r.id, &r.date, &r.data)
+				if err != nil {
+					logger.FromContext(ctx).Warn("Error retrieving information from database. Error: ", zap.Error(err))
+					return
+				}
+				m[r.id] = r
+			}
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/db_test.go:140]: nitpick: I think this part belongs to `RetrieveDatabaseInformation`

**Review by sven-urbanski-freiheit-com — CHANGES_REQUESTED** [review]
.

File: go.mod [hunk 7]
```diff
@@ -172,6 +173,7 @@ require (
 	github.com/lestrrat-go/httprc v1.0.5 // indirect
 	github.com/lestrrat-go/iter v1.0.2 // indirect
 	github.com/lestrrat-go/option v1.0.1 // indirect
+	github.com/mattn/go-sqlite3 v1.14.16 // indirect
```
[the remark this answers is not in the document]
Comment (sven-urbanski-freiheit-com) [reply go.mod:176]: mattn uses CGO, that's why I wanted to go with ncruces.

[the remark this answers is not in the document]
Comment (miguel-crespo-fdc) [reply go.mod:176]: It appears that this is imported by the migration library I used to run the migrations

File: services/cd-service/pkg/repository/db.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+	result, err := db.Exec("INSERT INTO dummy_table (id , created , data)  VALUES (?, ?, ?);", rand.Intn(9999), time.Now(), message)
+
+	if err != nil {
+		return nil, fmt.Errorf("Error inserting information into DB. Error: %w\n", err)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("Error querying the database. Error: %w\n", err)
+	}
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/db.go:126]: this is dead code, please remove

**Review by sven-urbanski-freiheit-com — APPROVED** [review]

**Pull request merged**